### ISSUE TITLE
virtual functions should specify exactly one of `virtual`, `override`, or `final`

### ISSUE THREAD

**Issue by zygoloid** [isocpp/CppCoreGuidelines#423]
[Suggested extension to C.128]

Using both `virtual` and `final` in the same declaration is error-prone:

    struct Base {
      virtual void f() const;
    };
    struct Derived {
      virtual void f() final;
    };

Note that `Derived::f()` does *not* override `Base::f()`, and is permitted despite being `final` and overriding nothing. If we follow C.128 literally (all overrides must say `override`) then we find that we must write this:

    virtual void f() final override;

... but that is verbose in the extreme. The same semantics can be obtained by the much simpler declaration:

    void f() final;

Likewise, `virtual override` is unnecessarily verbose, and the `virtual` keyword serves no purpose to either the reader or the compiler.

Thus we can conclude that all combinations with more than one of `virtual`, `override`, or `final` are needlessly verbose and possibly also error-prone, and can derive the following rule, which supersedes C.128:

> Use exactly one of `virtual`, `final`, and `override` when declaring a virtual function. Only use `virtual` when declaring a new virtual function.
>
> **Enforcement:**
>  * Compare names in base and derived classes and flag uses of the same name that do not override.
>  * Flag overrides with neither `override` nor `final`.
>  * Flag declarations with more than one of `virtual`, `override`, or `final`.

**Comment by hsutter** [comment]
Seems reasonable – I tend to prefer to write only one of the words.

 

Just confirming re “final” though, writing “final” implies “override,” so that “final” without “virtual” is illegal unless there is a base function being overridden? If you have the clause reference handy where I can find the rule that would be helpful, I couldn’t find it in a quick search.

**Comment by zygoloid** [comment]
`final` requires that the function be `virtual`, but it does not require it to override a function from a base class. In particular, this is valid:

    struct X { virtual void f() final {} };

But this is not:

    struct X { void f() final {} }; // error, final function 'f' is not virtual

The relevant rules are [class.virtual](10.3)/4 and /5, and [class.mem] (9.2)/8.

**Comment by AndrewPardoe** [comment]
Editors discussed 8 Dec 2015.

**Comment by hsutter** [comment]
Discussed on 12-8, we like the suggestion. Thanks, Richard!

**Comment by tituswinters** [comment]
(Google has generally liked this rule.)

In general there's been exactly one case where we've needed to make exceptions for this. When a virtual interface (with numerous derived implementations) is changed, you'll likely implement the new interface in the parent class, get the sub-classes to implement the new interface, convert callers, and then start deleting the old.  At one point in there, it's useful to mark the interface in the parent as virtual+final to prevent new implementations of the old interface.  

Otherwise, it's extremely rare to need anything else.  

The other thing that we've found to be surprising is that it's *technically* correct to mark sub-class destructors as "override" but most people find that confusing and naively prefer virtual.  (It might be worth adding a note suggesting what you believe to be proper guidance on those points).
